shared/mad/codeql/mad/modelgenerator/internal/ModelGeneratorImpl.qll

```diff
@@ -393,177 +366,302 @@ module MakeModelGenerator<
+     * return value or parameter or if `api` is a fluent API.
+     */
+    string captureFlow(DataFlowSummaryTargetApi api) {
+      result = captureQualifierFlow(api) or
+      result = captureThroughFlow(api)
+    }
+
+    /**
+     * Gets the neutral summary model for `api`, if any.
+     * A neutral summary model is generated, if we are not generating
+     * a summary model that applies to `api`.
+     */
+    string captureNoFlow(DataFlowSummaryTargetApi api) {
+      not exists(DataFlowSummaryTargetApi api0 |
+        exists(captureFlow(api0)) and api0.lift() = api.lift()
+      ) and
+      api.isRelevant() and
+      result = ModelPrinting::asNeutralSummaryModel(api)
     }
 
-    DataFlow::FlowFeature getAFeature() {
-      result instanceof DataFlow::FeatureEqualSourceSinkCallContext
+    /**
+     * A data flow configuration used for finding new sources.
+     * The sources are the already known existing sources and the sinks are the API return nodes.
+     *
+     * This can be used to generate Source summaries for an API, if the API expose an already known source
+     * via its return (then the API itself becomes a source).
+     */
+    module PropagateFromSourceConfig implements DataFlow::ConfigSig {
+      predicate isSource(DataFlow::Node source) {
+        exists(string kind |
+          isRelevantSourceKind(kind) and
+          sourceNode(source, kind)
+        )
+      }
+
+      predicate isSink(DataFlow::Node sink) {
+        sink instanceof ReturnNodeExt and
+        getEnclosingCallable(sink) instanceof DataFlowSourceTargetApi
+      }
+
+      DataFlow::FlowFeature getAFeature() { result instanceof DataFlow::FeatureHasSinkCallContext }
+
+      predicate isBarrier(DataFlow::Node n) {
+        exists(Type t | t = n.(NodeExtended).getType() and not isRelevantType(t))
+      }
+
+      predicate isAdditionalFlowStep(DataFlow::Node node1, DataFlow::Node node2) {
+        isRelevantTaintStep(node1, node2)
+      }
     }
-  }
 
-  module PropagateFlow = TaintTracking::GlobalWithState<PropagateFlowConfig>;
+    private module PropagateFromSource = TaintTracking::Global<PropagateFromSourceConfig>;
 
-  /**
-   * Gets the summary model(s) of `api`, if there is flow from parameters to return value or parameter.
-   */
-  string captureThroughFlow0(
-    DataFlowSummaryTargetApi api, DataFlow::ParameterNode p, ReturnNodeExt returnNodeExt
-  ) {
-    exists(string input, string output |
-      getEnclosingCallable(p) = api and
-      getEnclosingCallable(returnNodeExt) = api and
-      input = parameterNodeAsInput(p) and
-      output = getOutput(returnNodeExt) and
-      input != output and
-      result = ModelPrinting::asLiftedTaintModel(api, input, output)
-    )
-  }
+    /**
+     * Gets the source model(s) of `api`, if there is flow from an existing known source to the return of `api`.
+     */
+    string captureSource(DataFlowSourceTargetApi api) {
```

Comment:
Do `captureSource` and `captureSink` also use heuristics? I was under the impression they just used normal data flow?

Reply:
The dataflow configuration for both of these implements the `isAdditionalFlowStep` predicate like
```
predicate isAdditionalFlowStep(DataFlow::Node node1, DataFlow::Node node2) {
  isRelevantTaintStep(node1, node2)
}
```
The `isRelevantTaintStep` includes some read/store logic for fields which causes field conflation. In any case for some languages we can't avoid this, as declaring a field of an Argument or a ReturnValue as a source or sink doesn't work (this is not supported for C#).